Implement Biomes & StructureLayer Fix

aesliva opened the pull request:
Changed NeighborCountMax and NeighborCount to bytes (will only ever have 6 max). Seemed that Neighbors didn't need to be unsubscribed everytime chunk was resetted. 2nd condition in if statements in UpdateNeighborCount() didn't seem necessary (I could be wrong).

aesliva committed:
Optimize Chunk.UpdateNeighborCount() and Chunk.Reset()

Changed NeighborCountMax and NeighborCount to bytes (will only ever have 6 max). Seemed that Neighbors didn't need to be unsubscribed everytime chunk was resetted. 2nd condition in if statements in UpdateNeighborCount() didn't seem necessary (I could be wrong).

aesliva committed:
Implement Biomes

May not be the best way to go about doing this.. but it works. Would like to hear your opinion!

aesliva committed:
Revert "Implement Biomes"

This reverts commit 8976f981207adb29a9d57028e99d4972cb64f2e5.

richardbiely commented on VoxeUnity/Assets/Voxelmetric/Code/Core/Chunk.cs:
```diff
@@ -52,9 +52,9 @@ public partial class Chunk : ChunkEventSource
+        public byte NeighborCount { get; private set; }
         //! Maximum possible number of neighbors given the circumstances
-        public int NeighborCountMax { get; private set; }
+        public byte NeighborCountMax { get; private set; }
```
While this might be an optimization of sorts, I don't believe it's necessary at this point. Even with 10.000 chunks, we're talking about saving ~60 kiB of memory here.

richardbiely commented on VoxeUnity/Assets/Voxelmetric/Code/Core/Chunk.cs:
```diff
@@ -427,21 +431,21 @@ private static void UpdateNeighborCount(Chunk chunk)
+            byte maxNeighbors = 0;
             Vector3Int pos = chunk.Pos;
-            if (world.CheckInsideWorld(pos.Add(Env.ChunkSize, 0, 0)) && pos.x <= world.Bounds.maxX)
+            if (world.CheckInsideWorld(pos.Add(Env.ChunkSize, 0, 0)))
```
These conditions are necessary.
Based on your settings you might want to have chunks at the visible world bounds created.

aesliva replied:
Ah, I see.

richardbiely reviewed (commented):
I really appreciate your efforts with this one.
However, in this specific case, I find the proposed changes not really helping anything (int->byte) or introducing a bug to the code (removal of the second part of the conditions).
I'm sure you'll come with something else. There's like a ton of things to change/fix/improve in the code :)

richardbiely reviewed (commented):
I wonder why the revert.

aesliva commented:
Reverted because I added a layer type that is quite specific to my project (allows grass/flowers to spawn in groups and then have a falloff where they have decreasing chances of spawning as the noise value gets closer to 0, good for meadows!) Just an effort to keep things generalized, as other people may have other purposes. :) I just forgot to create another pull request!

aesliva committed:
Revert "Revert "Implement Biomes""

This reverts commit 325493e0a6e6964189d216e4c318c93b9a489a4f.

aesliva committed:
Fix for prev commit

aesliva committed:
Biome fix

aesliva committed:
Don't call GeneratedStructure.Build for each chunk on y-axis

X and Z axis was being used to see where to generate a tree. Y axis was ignored therefor Build was being called for each chunk on the y axis. The if statement fixed this problem, and improved performance.

richardbiely reviewed (changes requested):
Could you please create a separate merge request for this one, please?

aesliva commented:
Of course! My apologies, I'm new to git.

richardbiely commented:
Moved to a different pull request.
